custom vgg16 implementation

tristaaan opened the pull request:
This might be more accurate to the paper, however we cannot carry over the vgg16 weights which we were bfore. Despite this, the model performs almost identically!?

tristaaan committed:
custom vgg16 implementation

This might be more accurate to the paper, however we cannot carry over the vgg16 weights which we were bfore. Despite this, the model performs almost identically !?

tristaaan committed:
shuffle samples, zero padding layer

tristaaan commented:
I think it's really mysterious how we can't seem to get under 0.026 and it always early stops in under 20 epochs. I really think it's my training size now.

tristaaan commented:
```
Total             12178 # train and validation are this size
Training images   380 # total / 32
Validation images 380 # total / 32
```

So where are the other 10k or so images?

tristaaan commented:
in `data_preprocessor` is this the right split? 

```
data_size = len(img_paths)
train_size      = int(data_size * 0.4)
validation_size = int(data_size * 0.4)
test_size       = int(data_size * 0.2)
```

Why do we care about test set here?

tristaaan commented:
( @diegocasmo @ankurshukla03 )

ankurshukla03 commented:
Shouldn't in data_preprocess we should just split the data in training and validation ?
80-20%?

tristaaan commented:
We shouldn't even be doing like this unless we care about reproducing results. I'm going to be sampling from the file list from now on with 80-20 split

tristaaan commented:
mmmmmm there it is 
```
Traning images:   49262
Validtn images:   12316
Training batch:   1539
Validation batch: 384
```
Epochs take 151 seconds now.

ankurshukla03 commented:
so you are training on whole dataset.
kind of 2 hours if training for 50 epochs

ankurshukla03 commented:
We still need to fill the weekly plan

tristaaan commented:
> so you are training on whole dataset.
> kind of 2 hours if training for 50 epochs

Seems so! Sticking around for an early stop, I'm not trusting it until I see some loss under .026 or we're past 20 epochs

tristaaan committed:
Added test submission cells

tristaaan committed:
fixed path typo

tristaaan committed:
re organize channels

tristaaan commented on code/custom-vgg16-gpu.ipynb:
```diff
@@ -143,54 +151,54 @@
-      "Total params: 50,526,053\n",
-      "Trainable params: 50,526,053\n",
+      "Total params: 39,194,405\n",
+      "Trainable params: 39,194,405\n",
```
By reorganizing the input channels we require significantly fewer network parameters. This number here is down from 50.5 million where the input channel was organized 106,106,3. Here it's 3,106,106

tristaaan committed:
less patience, zero-padding layers

tristaaan committed:
different batching, no imrovements?

tristaaan commented:
The graph looks a bit better now, batches I believe are the right size and implemented properly. I’m really not sure why the batches are the size they are in the reference notebook (they’re 1?). In custom-vgg (no batch fix) I uploaded another solution csv last night and we’re performing a tad worse than the central pixel benchmark. So there is some progress I think it’s a good start.

ankurshukla03 committed:
debug

tristaaan committed:
full run.

This ran in approximately 12 hours. I accidently reran the CNN cell so the history is gone, but I have the weights saved and the history graph is correct.

tristaaan committed:
submission generation fix

tristaaan commented:
It ran for the full 50 epochs, and we got to a loss of 0.1697. We can keep training this with the weights it generated if desired. It took approximately 12 hours to train which I think can be reduced a bit by fully preprocessing the images before hand so that the GPU is always working.

I ran the predictions through Kaggle and we now score **0.13070**. This would have put us at **153rd**. This is also better than the reference notebook we're basing this off of.

Repo-wise, I'm not quite sure how to reconcile the batch changes (move them to custom vgg and delete the batch-fix notebook?) I'm saving the weights interdependently so we can refer back to this model or train it more if desired.

ankurshukla03 commented:
Amazing, now we really got a good score.

> I ran the predictions through Kaggle and we now score 0.13070. This would have put us at 153rd. This is also better than the reference notebook we're basing this off of.

diegocasmo reviewed (commented):
- As of now, we should only have 2 notebooks. The one that uses the pre-processed dataset to deal with less images (so that we can tell how parameters affect training without the need to use the whole dataset and submit to Kaggle), and then the notebook which uses the entire dataset and creates the submission ``.csv``
- The ``def preprocess_image(paths, i)`` should instead be ``def preprocess_image(img_path)``. No need for that method to be aware there's an image path array and find a particular index on it
- How does ``batch_size`` affect training? While ``32`` is suggested almost everywhere, what about if its ``16``, or ``64``? I expect the epoch time to be reduced/increased accordingly, but I'm most interested in checking whether it helps the weight updates or not (this might be easier to test in the notebook which does not use the entire dataset). Feel free to do this in another branch if desired.
  -  From the winner: ``I used a minibatch size of 16 examples.``
- This is the data pre-processing done by the winner of the competiton: ``[...] so I cropped all images to 207x207. I then downsampled them 3x to 69x69 [...]``. I wonder if we should change ours to match his?
- About training and validation set sizes (from the winner as well): ``For validation purposes, I split the training set in two parts. I used the first 90% for training, and the remainder for validation. ``
- Images are still not being rotated by a random multiple of  ``90`` (not sure how much this would help, but again, might be worth trying it in the notebook which uses the smaller dataset)
- Finally, for the sake of reproducibility, we should remove the randomization from the data processor notebook, that way we always train/validate/test with the same small dataset. This will allows us to better understand the hyper-parameters effect on learning.

tristaaan commented:
> This is the data pre-processing done by the winner of the competition: [...] so I cropped all images to 207x207. I then downsampled them 3x to 69x69 [...]. I wonder if we should change ours to match his?

Wow, even more dimensionality reduction!

> Finally, for the sake of reproducibility, we should remove the randomization from the data processor notebook, that way we always train/validate/test with the same small dataset. This will allows us to better understand the hyper-parameters effect on learning.

Set a random seed before performing the split.

> From the winner: I used a minibatch size of 16 examples.

Let's do a run with batches of 16 then (#25 )

tristaaan committed:
removed old notebooks, renamed latest vgg16 notebook, batch-preprocessor

tristaaan committed:
fix #26 better data splitting

tristaaan committed:
moved testing cells to new file

tristaaan committed:
clean up and comments

tristaaan commented:
I'm thinking of altering this a bit: 

> As of now, we should only have 2 notebooks. [...]

to 

- [x] vgg script with a parameter to set the dataset size either as a fraction of the whole dataset or a fixed subset.
- [x] submission script

> The def preprocess_image(paths, i) should instead be def preprocess_image(img_path). No need for that method to be aware there's an image path array and find a particular index on it

This has been cleared up too. Amazing the code you can write when just fumbling around.

diegocasmo commented:
@tristaaan 

Can you export your conda environment too? ``.yml`` file seems outdated

tristaaan committed:
updated environment file

tristaaan commented:
I exported what I'm using on my PC, it looks pretty different from the original one, but as long as conda can handle it. 

Caveat: I'm using `tensorflow-gpu` which is different from `tensorflow`

tristaaan committed:
variable or fixed training set size, sample seed

tristaaan committed:
split documentation, image batcher options

tristaaan committed:
automatically adjust model input based on training image shape

tristaaan commented:
PTAL, some things are pretty reorganized.

diegocasmo reviewed (approved):
Have you trained with the newly preprocessed data? what result do we get (Kaggle)?

tristaaan commented:
> Have you trained with the newly preprocessed data? what result do we get (Kaggle)?

Not yet, wanted to get this in order first. I'll continue this comment in #25
